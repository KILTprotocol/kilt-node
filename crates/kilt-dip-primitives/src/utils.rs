// KILT Blockchain â€“ https://botlabs.org
// Copyright (C) 2019-2024 BOTLabs GmbH

// The KILT Blockchain is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// The KILT Blockchain is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

// If you feel like getting in touch with us, you can do so at info@botlabs.org

/// The output of a type implementing the [`sp_runtime::traits::Hash`] trait.
pub type OutputOf<Hasher> = <Hasher as sp_runtime::traits::Hash>::Output;

pub(crate) use calculate_parachain_head_storage_key::*;
mod calculate_parachain_head_storage_key {
	use parity_scale_codec::Encode;
	use sp_core::storage::StorageKey;

	pub(crate) fn calculate_parachain_head_storage_key(para_id: u32) -> StorageKey {
		StorageKey(
			[
				frame_support::storage::storage_prefix(b"Paras", b"Heads").as_slice(),
				sp_io::hashing::twox_64(para_id.encode().as_ref()).as_slice(),
				para_id.encode().as_slice(),
			]
			.concat(),
		)
	}

	#[cfg(test)]
	mod tests {
		use super::*;

		#[test]
		fn calculate_parachain_head_storage_key_successful_spiritnet_parachain() {
			assert_eq!(
				calculate_parachain_head_storage_key(2_086).0,
				hex_literal::hex!(
					"cd710b30bd2eab0352ddcc26417aa1941b3c252fcb29d88eff4f3de5de4476c32c0cfd6c23b92a7826080000"
				)
				.to_vec()
			);
		}
		#[test]
		fn calculate_parachain_head_storage_key_successful_peregrine_parachain() {
			assert_eq!(
				calculate_parachain_head_storage_key(2_000).0,
				hex_literal::hex!(
					"cd710b30bd2eab0352ddcc26417aa1941b3c252fcb29d88eff4f3de5de4476c363f5a4efb16ffa83d0070000"
				)
				.to_vec()
			);
		}
	}
}

pub(crate) use calculate_dip_identity_commitment_storage_key_for_runtime::*;
mod calculate_dip_identity_commitment_storage_key_for_runtime {
	use pallet_dip_provider::IdentityCommitmentVersion;
	use sp_core::storage::StorageKey;

	pub(crate) fn calculate_dip_identity_commitment_storage_key_for_runtime<Runtime>(
		subject: &Runtime::Identifier,
		version: IdentityCommitmentVersion,
	) -> StorageKey
	where
		Runtime: pallet_dip_provider::Config,
	{
		StorageKey(pallet_dip_provider::IdentityCommitments::<Runtime>::hashed_key_for(
			subject, version,
		))
	}

	#[cfg(test)]
	mod tests {
		use super::*;

		#[test]
		fn calculate_dip_identity_commitment_storage_key_for_runtime_successful_peregrine_parachain() {
			use did::DidIdentifierOf;
			use peregrine_runtime::Runtime as PeregrineRuntime;
			use sp_core::crypto::Ss58Codec;

			assert_eq!(
			calculate_dip_identity_commitment_storage_key_for_runtime::<PeregrineRuntime>(&DidIdentifierOf::<PeregrineRuntime>::from_ss58check("4s3jpR7pzrUdhVUqHHdWoBN6oNQHBC7WRo7zsXdjAzQPT7Cf").unwrap(), 0).0,
			hex_literal::hex!("b375edf06348b4330d1e88564111cb3d5bf19e4ed2927982e234d989e812f3f314c9211b34c8b43b2a18d67d5c96de9cb6caebbe9e3adeaaf693a2d198f2881d0b504fc72ed4ac0a7ed24a025fc228ce01a12dfa1fa4ab9a0000")
				.to_vec()
		);
		}
	}
}
