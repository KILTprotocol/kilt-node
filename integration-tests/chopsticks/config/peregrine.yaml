endpoint:
  - wss://peregrine.kilt.io
mock-signature-host: true
db: db/peregrine.sqlite
runtime-log-level: 5
wasm-override: /home/antonio/Developer/kilt-node/target/debug/wbuild/peregrine-runtime/peregrine_runtime.compact.compressed.wasm
