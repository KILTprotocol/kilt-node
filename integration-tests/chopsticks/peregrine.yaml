db: './db/peregrine.db.sqlite'
port: 9004
endpoint: 'wss://peregrine.kilt.io'
wasm-override: ./peregrine_runtime.compact.compressed.wasm

import-storage:
    assetSwitchPool1:
        switchPair:
            # Pool account id Charlie
            poolAccountId: '5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y'
            remoteAssetBalance: '100000000000000000'
            remoteAssetSovereignTotalBalance: '90000000000000000000'
            remoteAssetCirculatingSupply: '10000000000000000000'
            remoteAssetTotalSupply: '100000000000000000000'
            remoteAssetId:
                V3:
                    Concrete:
                        parents: 2
                        interior:
                            X2:
                                - GlobalConsensus:
                                      Ethereum:
                                          chainId: 11155111
                                - AccountKey20:
                                      network: null
                                      key: '0x06012c8cf97bead5deae237070f9587f8e7a266d'
            remoteXcmFee:
                V3:
                    id:
                        concrete:
                            parents: 1
                            interior: 'Here'
                    fun:
                        Fungible: 1000000000000
            remoteReserveLocation:
                V3:
                    parents: 1
                    interior:
                        X1:
                            Parachain:
                                id: 1000
            status: 'Running'

    fungibles:
        asset:
            - - - parents: 1
                  interior: 'Here'
              # Alice account
              - owner: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                issuer: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                admin: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                freezer: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                supply: '4242424242424242424242'
                deposit: 100000000000
                minBalance: 100
                isSufficient: false
                accounts: 1
                sufficients: 0
                approvals: 0
                status: 'Live'

        account:
            - - - parents: 1
                  interior: 'Here'
                  # Alice account
                - '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
              - balance: '10000000000000000'
                status: 'Liquid'
                reason: 'Consumer'
                extra: null

    System:
        Account:
            # Alice account
            - - - '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
              - providers: 1
                data:
                    free: '100000000000000000'
            # Charlie account
            - - - '5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y'
              - providers: 1
                data:
                    free: '100000000000000000'
    polkadotXcm:
        safeXcmVersion: 3
