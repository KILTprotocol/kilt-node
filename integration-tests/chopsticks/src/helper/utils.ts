import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
