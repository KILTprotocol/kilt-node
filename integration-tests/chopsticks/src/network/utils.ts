export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
