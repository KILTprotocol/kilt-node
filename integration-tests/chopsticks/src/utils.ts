import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
