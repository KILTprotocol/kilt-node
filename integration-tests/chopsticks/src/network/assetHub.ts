import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'

import type { Config } from './types.js'
import { initialBalanceDOT, toNumber } from '../utils.js'

/// Options used to create the Spiritnet context
export const getSetupOptions = ({
	blockNumber = undefined,
	wasmOverride = undefined,
}: {
	blockNumber?: number
	wasmOverride?: string
}) =>
	({
		endpoint: process.env.ASSETHUB_WSS || 'wss://asset-hub-polkadot-rpc.dwellir.com',
		db: './db/assethub.db.sqlite',
		port: toNumber(process.env.ASSETHUB_PORT) || 9003,
		wasmOverride,
		blockNumber,
	}) as SetupOption

/// AssetHub has no own coin. Teleported dots are used as the native token.
export function assignDotTokensToAccountsAsStorage(addr: string[], balance: bigint = initialBalanceDOT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance.toString() } }]),
		},
	}
}

export function createForeignAsset(manager: string, assetId = eKiltLocation) {
	return {
		foreignAssets: {
			asset: [
				[
					[assetId],
					{
						owner: manager,
						issuer: manager,
						admin: manager,
						freezer: manager,
						// Just make it big enough
						supply: '10000000000000000000000000000',
						deposit: 0,
						minBalance: 0,
						isSufficient: false,
						accounts: 0,
						sufficients: 0,
						approvals: 0,
						status: 'Live',
					},
				],
			],
		},
	}
}

/// Assigns KSM to an account
export function assignKSMtoAccounts(addr: string[], balance: bigint = initialBalanceDOT) {
	return {
		foreignAssets: {
			account: addr.map((addr) => [
				[KSMAssetLocation, addr],
				{
					balance: balance,
					status: 'Liquid',
					reason: 'Consumer',
					extra: null,
				},
			]),
		},
	}
}

/// Assigns the foreign asset to the accounts.
/// Does not check if supply is matching the sum of the account balances.
export function assignForeignAssetToAccounts(accountInfo: [string, bigint][], assetId = eKiltLocation) {
	return {
		foreignAssets: {
			account: accountInfo.map(([account, balance]) => [
				[assetId, account],
				{
					balance: balance,
					status: 'Liquid',
					reason: 'Consumer',
					extra: null,
				},
			]),
		},
	}
}

/// AssetHub ParaId
export const paraId = 1000

export const KSMAssetLocation = {
	parents: 2,
	interior: {
		X1: {
			GlobalConsensus: 'Kusama',
		},
	},
}

// Sibling Sovereign Account
export const sovereignAccountOnSiblingChains = '4qXPdpimHh8TR24RSk994yVzxx4TLfvKj5i1qH5puvWmfAqy'

/// Native token in AssetHub
export const nativeTokenLocation = { parents: 1, interior: 'Here' }

export const eKiltLocation = {
	parents: 2,
	interior: {
		X2: [
			{
				GlobalConsensus: { Ethereum: { chainId: 11155111 } },
			},
			{
				AccountKey20: {
					network: null,
					key: '0x06012c8cf97bead5deae237070f9587f8e7a266d',
				},
			},
		],
	},
}

export async function getContext(): Promise<Config> {
	const options = getSetupOptions({})
	return setupContext(options)
}
