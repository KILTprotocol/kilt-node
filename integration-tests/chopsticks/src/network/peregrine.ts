import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'

import type { Config } from './types.js'
import { ROC, initialBalanceKILT, initialBalanceROC, toNumber } from '../utils.js'
import { AssetSwitchSupplyParameters } from '../types.js'

/// Options used to create the Peregrine context
const options = {
	endpoint: process.env.PEREGRINE_WS || 'wss://peregrine.kilt.io',
	db: './db/peregrine.db.sqlite',
	port: toNumber(process.env.PEREGRINE_PORT) || 9004,
	wasmOverride: process.env.PEREGRINE_WASM_OVERRIDE || undefined,
} as SetupOption

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

export function setSafeXcmVersion4() {
	return {
		polkadotXcm: {
			safeXcmVersion: 4,
		},
	}
}

export function createAndAssignRocs(manager: string, addr: string[], balance: bigint = initialBalanceROC) {
	return {
		fungibles: {
			asset: [
				[
					[ROC_LOCATION],
					{
						owner: manager,
						issuer: manager,
						admin: manager,
						freezer: manager,
						supply: balance * BigInt(addr.length),
						deposit: 0,
						minBalance: 1,
						isSufficient: false,
						accounts: addr.length,
						sufficients: 0,
						approvals: 0,
						status: 'Live',
					},
				],
			],
			account: addr.map((acc) => [
				[{ parents: 1, interior: 'here' }, acc],
				{ balance: balance, status: 'Liquid', reason: 'Consumer', extra: null },
			]),
		},
	}
}

export function setSudoKey(sudoKey: string) {
	return {
		Sudo: {
			key: sudoKey,
		},
	}
}

export function setSwitchPair(
	parameters: AssetSwitchSupplyParameters,
	remoteAssetId: Record<string, unknown>,
	remoteXcmFee: Record<string, unknown>,
	remoteReserveLocation: Record<string, unknown>,
	poolAccountId: string = initialPoolAccountId,
	status: 'Running' | 'Paused' = 'Running'
) {
	return {
		assetSwitchPool1: {
			SwitchPair: {
				poolAccountId,
				remoteAssetSovereignTotalBalance: parameters.sovereignSupply,
				remoteAssetCirculatingSupply: parameters.circulatingSupply,
				remoteAssetTotalSupply: parameters.totalSupply,
				remoteAssetId,
				remoteXcmFee,
				remoteReserveLocation,
				status,
			},
		},
		// the pool account needs at least as much fund to cover the circulating supply. Give him exactly that amount + ED.
		System: {
			Account: [[[poolAccountId], { providers: 1, data: { free: parameters.circulatingSupply } }]],
		},
	}
}

/// Peregrine ParaId
export const paraId = 2086
export const PILT = { Concrete: { parents: 0, interior: 'Here' } }
export const ROC_LOCATION = {
	parents: 1,
	interior: 'Here',
}
// 0.1 ROC as remote fee
export const remoteFee = ROC / BigInt(10)

/// Sibling sovereign account for other chains
export const sovereignAccountAsSibling = '5Eg2fnshxV9kofpcNEFE7azHLAjcCtpNkbsH3kkWZasYUVKs'
// ED on Peregrine
export const existentialDeposit = BigInt('10000000000000')

export const initialPoolAccountId = '4nv4phaKc4EcwENdRERuMF79ZSSB5xvnAk3zNySSbVbXhSwS'

export async function getContext(): Promise<Config> {
	return setupContext(options)
}

export const treasuryAccount = '5EYCAe5hitmhtZcXaCrygrqw4kbmJ8bGm15PBuwUdDdQTUgY'
export const parachainStakingRewards = BigInt('760514107536694')
