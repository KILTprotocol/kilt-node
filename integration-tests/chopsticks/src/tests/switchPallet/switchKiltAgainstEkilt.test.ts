import { test } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import * as PeregrineConfig from '../../network/peregrine.js'
import * as AssetHubConfig from '../../network/assetHub.js'
import {
	getAssetSwitchParameters,
	initialBalanceKILT,
	initialBalanceROC,
	keysAlice,
	keysCharlie,
	ROC,
} from '../../utils.js'
import {
	peregrineContext,
	getFreeBalancePeregrine,
	getFreeRocPeregrine,
	getFreeEkiltAssetHub,
	assethubContext,
	getFreeRocAssetHub,
	getRemoteLockedSupply,
} from '../index.js'
import { checkSwitchPalletInvariant, checkBalance, createBlock, setStorage, hexAddress } from '../utils.js'
import { getAccountLocationV4, getSiblingLocationV4 } from '../../network/utils.js'

test('Switch PILTs against ePILTS on AssetHub', async ({ expect }) => {
	const { checkEvents } = withExpect(expect)

	const switchParameters = getAssetSwitchParameters()
	const feeAmount = (ROC * BigInt(10)) / BigInt(100)

	const remoteAssetId = { V4: AssetHubConfig.eKiltLocation }
	const remoteXcmFeeId = { V4: { id: AssetHubConfig.nativeTokenLocation, fun: { Fungible: feeAmount } } }
	const remoteReserveLocation = getSiblingLocationV4(AssetHubConfig.paraId)

	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address], initialBalanceROC),
		...PeregrineConfig.setSafeXcmVersion4(),
	})

	await setStorage(
		peregrineContext,
		PeregrineConfig.setSwitchPair(switchParameters, remoteAssetId, remoteXcmFeeId, remoteReserveLocation)
	)

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccountsAsStorage(
			[keysAlice.address, PeregrineConfig.sovereignAccountAsSibling],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address),
	})

	await setStorage(
		assethubContext,
		AssetHubConfig.assignForeignAssetToAccounts([
			[PeregrineConfig.sovereignAccountAsSibling, switchParameters.sovereignSupply],
		])
	)

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// initial balance of the pool account and sovereign account
	const initialBalancePoolAccount = await getFreeBalancePeregrine(PeregrineConfig.initialPoolAccountId)
	const initialBalanceSovereignAccount = await getFreeEkiltAssetHub(PeregrineConfig.sovereignAccountAsSibling)
	const initialBalanceRocSovereignAccount = await getFreeRocAssetHub(PeregrineConfig.sovereignAccountAsSibling)
	const initialRemoteLockedSupply = await getRemoteLockedSupply()

	// 50 PILTS
	const balanceToTransfer = BigInt('50000000000000000')

	const beneficiary = getAccountLocationV4(hexAddress(keysAlice.address))

	const signedTx = peregrineContext.api.tx.assetSwitchPool1.switch(balanceToTransfer.toString(), beneficiary)
	const fees = (await signedTx.paymentInfo(keysAlice)).partialFee.toBigInt()

	const events = await sendTransaction(signedTx.signAsync(keysAlice))

	await createBlock(peregrineContext)

	await checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender Peregrine::xcmpQueue::[XcmpMessageSent]')
	await checkEvents(events, 'assetSwitchPool1').toMatchSnapshot(
		'sender Peregrine::assetSwitchPool1::[LocalToRemoteSwitchExecuted]'
	)
	await checkEvents(events, { section: 'balances', method: 'Transfer' }).toMatchSnapshot(
		'sender Peregrine::balances::[Transfer]'
	)

	// check balance. Alice had 50 PILts
	const freeBalanceAlice = await getFreeBalancePeregrine(keysAlice.address)
	expect(freeBalanceAlice).toBe(initialBalanceKILT - balanceToTransfer - fees)

	// check balance Alice. Some fees should have been paid with her rocs:
	const freeRocBalanceAlice = await getFreeRocPeregrine(keysAlice.address)
	expect(freeRocBalanceAlice).eq(initialBalanceROC - BigInt(PeregrineConfig.remoteFee))

	// the Switch pool account should have 50 more PILTs
	const balancePoolAccountAfterTx = await getFreeBalancePeregrine(PeregrineConfig.initialPoolAccountId)
	expect(balancePoolAccountAfterTx).eq(initialBalancePoolAccount + balanceToTransfer)

	// process the msg
	await createBlock(assethubContext)

	// alice should have the exact transferred amount of eKILT. Fees are paid by sovereign account
	const freeBalanceAliceAssetHub = await getFreeEkiltAssetHub(keysAlice.address)
	expect(freeBalanceAliceAssetHub).eq(balanceToTransfer)

	// sovereign account should have less eKILT by the amount of the transferred PILTs
	const freeBalanceSovereignAccount = await getFreeEkiltAssetHub(PeregrineConfig.sovereignAccountAsSibling)
	expect(initialBalanceSovereignAccount - balanceToTransfer).eq(freeBalanceSovereignAccount)

	// sovereign account should have paid the fees. Calculating the fees is not simple in context of XCM.
	// We just check that the balance has decreased
	const freeRocsSovereignAccount = await getFreeRocAssetHub(PeregrineConfig.sovereignAccountAsSibling)
	expect(freeRocsSovereignAccount).toBeLessThan(initialBalanceRocSovereignAccount)

	// remote locked supply should have decreased by the amount of the transferred PILTs
	const remoteLockedSupply = await getRemoteLockedSupply()
	expect(remoteLockedSupply).eq(initialRemoteLockedSupply - balanceToTransfer)

	await checkSwitchPalletInvariant(expect)
}, 20_000)
