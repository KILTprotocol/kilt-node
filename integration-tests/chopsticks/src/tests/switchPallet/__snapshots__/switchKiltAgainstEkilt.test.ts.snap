// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch PILTs against ePILTS on AssetHub > sender Peregrine::assetSwitchPool1::[LocalToRemoteSwitchExecuted] 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": {
        "V4": {
          "interior": {
            "X1": [
              {
                "AccountId32": {
                  "id": "(hash)",
                  "network": null,
                },
              },
            ],
          },
          "parents": 0,
        },
      },
    },
    "method": "LocalToRemoteSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Switch PILTs against ePILTS on AssetHub > sender Peregrine::balances::[Transfer] 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": "4nv4phaKc4EcwENdRERuMF79ZSSB5xvnAk3zNySSbVbXhSwS",
    },
    "method": "Transfer",
    "section": "balances",
  },
]
`;

exports[`Switch PILTs against ePILTS on AssetHub > sender Peregrine::xcmpQueue::[XcmpMessageSent] 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
